nbgitautosync/static/index.js

```diff
@@ -0,0 +1,303 @@
+require([
+    'jquery',
+    'base/js/utils',
+    'components/xterm.js/dist/xterm',
+], function(
+    $,
+    utils,
+    Terminal
+) {
+
+
+    function GitSync(baseUrl, repo, branch, path) {
+        // Class that talks to the API backend & emits events as appropriate
+        this.baseUrl = baseUrl;
+        this.repo = repo;
+        this.branch = branch;
+        this.path = path;
+
+        if (path.endsWith('.ipynb')) {
+            this.redirectUrl = baseUrl + 'notebooks/' + path;
+        } else {
+            this.redirectUrl = baseUrl + 'tree/' + path;
+        }
+
+        this.callbacks = {};
+    }
+
+    GitSync.prototype.addHandler = function(event, cb) {
+        if (this.callbacks[event] == undefined) {
+            this.callbacks[event] = [cb];
+        } else {
+            this.callbacks[event].push(cb);
+        }
+    };
+
+    GitSync.prototype._emit = function(event, data) {
+        if (this.callbacks[event] == undefined) { return; }
+        $.each(this.callbacks[event], function(i, ev) {
+            ev(data);
+        });
+    };
+
+
+    GitSync.prototype.start = function() {
+        var syncUrl = this.baseUrl + 'git-sync/api?' + $.param({
+            repo: this.repo,
+            branch: this.branch,
+            path: this.path
+        });
+
+        this.eventSource = new EventSource(syncUrl);
+        var that = this;
+        this.eventSource.addEventListener('message', function(ev) {
+            var data = JSON.parse(ev.data);
+            if (data.phase == 'finished' || data.phase == 'error') {
+                that.eventSource.close();
+            }
+            that._emit(data.phase, data);
+        });
+        this.eventSource.addEventListener('error', function(error) {
+            that._emit('error', error);
+        });
+    };
+
+    function GitSyncView(termElement, progressElement, termToggleElement) {
+        // Class that encapsulates view rendering as much as possible
+        // jQuery free, but not too strong an opinion about it.
+        this.term = new Terminal({
+            convertEol: true,
+            disableStdin: true
+        });
+        this.term.open(termElement);
+        this.progressElement = progressElement;
+
+        this.termToggleElement = termToggleElement;
+
+        var that = this;
+        this.termToggleElement.addEventListener('click', function() {
+            termElement.parentElement.classList.toggle('hidden');
+            that._fitTerm();
+        });
+    }
+
+
+    GitSyncView.prototype.setProgressValue = function(val) {
+        this.progressElement.setAttribute('aria-valuenow', val);
+        this.progressElement.style.width = val + '%';
+    };
+
+    GitSyncView.prototype.getProgressValue = function() {
+        return parseFloat(this.progressElement.getAttribute('aria-valuenow'));
+    };
+
+    GitSyncView.prototype.setProgressText = function(text) {
+        this.progressElement.getElementsByTagName('span')[0].innerText = text;
```

Comment:
Very minor, but would the jquery version of this be cleaner? ie. `$( this.progressElement ).find( "span" );`